### PR TITLE
feat(command): subscribe command request via internal messaging

### LINKED ISSUE #4170: Core command subscribes to internal command requests and publishes requests to the appropriate device service via messaging
# 🚀 Feature Request

### Relevant Package [**REQUIRED**]

 This feature request is for core-command

### Description [**REQUIRED**]
 A clear and concise description of the problem or missing capability...
https://docs.edgexfoundry.org/2.2/design/adr/0023-North-South-Messaging/#edgex-service-internal-message-bus-requests

### Describe the solution you'd like
 If you have a solution in mind, please describe it.

### Describe alternatives you've considered
 Have you considered any alternative solutions or workarounds?

### PR DESCRIPTION
**Pull request by hahattan**
core-command subscribes to internal command requests and publishes requests to the to the appropriate device service via messaging

Signed-off-by: Chris Hung <chris@iotechsys.com>

fix #4170

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions

1. Run core-command from this branch with `MESSAGEQUEUE_REQUIRED=true` environment variable
2. Send command requests via `redis` container to InternalRequestCommandTopic, for example:
```shell
/data # redis-cli
127.0.0.1:6379> PUBLISH .command.request.Random-Boolean-Device.Bool.get '{"CorrelationID": "14a42ea6-c394-41c3-8bcd-a29b9f5e6835", "ApiVersion": "v2", "RequestId": "e6e8a2f4-eb14-4649-9e2b-175247911369", "Content-Type": "application/json"}'
(integer) 1
```
3. Verified core-command receives the request, propagates to device service, receives the response and send back to internal messaging (`redis` container) on InternalResponseCommandTopicPrefix topic:
```
level=DEBUG ts=2022-10-03T05:53:17.983273178Z app=core-command source=internal.go:107 msg="Command request received on message queue. Topic: /command/request/Random-Boolean-Device/Bool/get, Correlation-id: 14a42ea6-c394-41c3-8bcd-a29b9f5e6835 "
level=DEBUG ts=2022-10-03T05:53:17.987350115Z app=core-command source=internal.go:147 msg="Command request sent to message queue. Topic: edgex/command/request/device-virtual/Random-Boolean-Device/Bool/get, Correlation-id: 14a42ea6-c394-41c3-8bcd-a29b9f5e6835"
level=DEBUG ts=2022-10-03T05:53:17.987990896Z app=core-command source=internal.go:52 msg="Command response received on message queue. Topic: edgex/command/response/device-virtual/Random-Boolean-Device/Bool/get, Correlation-id: 14a42ea6-c394-41c3-8bcd-a29b9f5e6835 "
level=DEBUG ts=2022-10-03T05:53:17.988207693Z app=core-command source=internal.go:70 msg="Published response message on topic '/command/response/Random-Boolean-Device/Bool/get'"
```
`redis`:
```shell
/data # redis-cli
127.0.0.1:6379> PSUBSCRIBE .command.response.*
Reading messages... (press Ctrl-C to quit)
1) "psubscribe"
2) ".command.response.*"
3) (integer) 1
1) "pmessage"
2) ".command.response.*"
3) ".command.response.Random-Boolean-Device.Bool.get"
4) "{\"ReceivedTopic\":\"edgex/command/response/device-virtual/Random-Boolean-Device/Bool/get\",\"CorrelationID\":\"14a42ea6-c394-41c3-8bcd-a29b9f5e6835\",\"ApiVersion\":\"v2\",\"RequestID\":\"e6e8a2f4-eb14-4649-9e2b-175247911369\",\"ErrorCode\":0,\"Payload\":\"eyJhcGlWZXJzaW9uIjoidjIiLCJzdGF0dXNDb2RlIjoyMDAsImV2ZW50Ijp7ImFwaVZlcnNpb24iOiJ2MiIsImlkIjoiZmI3OWUwZWItMTBkYi00NWRjLTkzM2MtMGFjNjlmZjk0MzNjIiwiZGV2aWNlTmFtZSI6IlJhbmRvbS1Cb29sZWFuLURldmljZSIsInByb2ZpbGVOYW1lIjoiUmFuZG9tLUJvb2xlYW4tRGV2aWNlIiwic291cmNlTmFtZSI6IkJvb2wiLCJvcmlnaW4iOjE2NjQ3NzY0ODA1NTkwOTU1MjYsInJlYWRpbmdzIjpbeyJpZCI6Ijk4MjU5MWRlLTNlNDYtNGI0YS04Mjc5LTg2NDNkODVhYTM3MSIsIm9yaWdpbiI6MTY2NDc3NjQ4MDU1OTA5NTUyNiwiZGV2aWNlTmFtZSI6IlJhbmRvbS1Cb29sZWFuLURldmljZSIsInJlc291cmNlTmFtZSI6IkJvb2wiLCJwcm9maWxlTmFtZSI6IlJhbmRvbS1Cb29sZWFuLURldmljZSIsInZhbHVlVHlwZSI6IkJvb2wiLCJ2YWx1ZSI6InRydWUifV19fQ==\",\"ContentType\":\"application/json\",\"QueryParams\":{}}"

```

### CODE REVIEW

File: internal/core/command/controller/messaging/internal.go [hunk 1]
```diff
@@ -58,7 +59,93 @@ func SubscribeCommandResponses(ctx context.Context, router MessagingRouter, dic
...
+				}
+				lc.Debugf("Published response message on topic '%s'", responseTopic)
+			}
+		}
+	}()
+
+	return nil
+}
+
+func SubscribeCommandRequests(ctx context.Context, router MessagingRouter, dic *di.Container) errors.EdgeX {
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/internal.go:78]: Add Godoc style comments for public functions.

File: internal/core/command/controller/messaging/internal.go [hunk 2]
```diff
@@ -58,7 +59,93 @@ func SubscribeCommandResponses(ctx context.Context, router MessagingRouter, dic
...
+				commandName := topicLevels[length-2]
+				method := topicLevels[length-1]
+				if !strings.EqualFold(method, "get") && !strings.EqualFold(method, "set") {
+					lc.Errorf("Unknown request method: %s, only 'get' or 'set' is allowed", method)
+					lc.Warn("Not publishing error message back due to insufficient information on response topic")
+					continue
+				}
+				internalResponseTopic := strings.Join([]string{messageBusInfo.Internal.Topics[InternalResponseCommandTopicPrefix], deviceName, commandName, method}, "/")
+
+				internalRequestTopic, err := validateRequestTopic(messageBusInfo.Internal.Topics[RequestTopicPrefix], deviceName, commandName, method, dic)
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/internal.go:128]:
```suggestion
				deviceRequestTopic, err := validateRequestTopic(messageBusInfo.Internal.Topics[RequestTopicPrefix], deviceName, commandName, method, dic)
```

File: internal/core/command/controller/messaging/internal.go [hunk 3]
```diff
@@ -58,7 +59,93 @@ func SubscribeCommandResponses(ctx context.Context, router MessagingRouter, dic
...
+					err = messageBus.Publish(responseEnvelope, internalResponseTopic)
+					if err != nil {
+						lc.Errorf("Could not publish to topic '%s': %s", internalResponseTopic, err.Error())
+					}
+
+					continue
+				}
+
+				// expected internal command request topic scheme: #/<device-service>/<device>/<command-name>/<method>
+				err = messageBus.Publish(requestEnvelope, internalRequestTopic)
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/internal.go:141]:
```suggestion
				err = messageBus.Publish(requestEnvelope, deviceRequestTopic)
```

File: internal/core/command/controller/messaging/internal.go [hunk 4]
```diff
@@ -58,7 +60,95 @@ func SubscribeCommandResponses(ctx context.Context, router MessagingRouter, dic
...
+				}
+				lc.Debugf("Published response message to internal MessageBus on topic '%s'", responseTopic)
+			}
+		}
+	}()
+
+	return nil
+}
+
+// SubscribeCommandRequests subscribes command requests from EdgeX service (e.g., Application Service)
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/internal.go:79]: Perfect!

File: internal/core/command/controller/messaging/router.go [hunk 5]
```diff
@@ -13,7 +13,7 @@ import (
 // MessagingRouter defines interface for Command Service to know
 // where to route the receiving device command response.
 type MessagingRouter interface {
-	// ResponseTopic returns the responseTopicPrefix by requestId, and a boolean value
+	// DeviceResponseTopic returns the responseTopicPrefix by requestId, and a boolean value
```
Comment (lenny-intel) [inline internal/core/command/controller/messaging/router.go:16]: I don't think you meant to change this one.
```suggestion
	// ResponseTopic returns the responseTopicPrefix by requestId, and a boolean value
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]
One minor comment tweak needed, otherwise looks great!

**Review by lenny-intel — APPROVED** [review]
LGTM

File: cmd/core-command/res/configuration.toml [hunk 6]
```diff
@@ -54,12 +54,12 @@ Type = "consul"
...
-    InternalRequestQueryTopic = "/commandquery/request"   # for subscribing to internal command query requests
-    InternalResponseQueryTopic = "/commandquery/response"   # for publishing reponsses back to internal service
+    DeviceRequestTopicPrefix = "edgex/command/request"    # for publishing requests to the device service; <device-service>/<device-name>/<command-name>/<method> will be added to this publish topic prefix
+    DeviceResponseTopic = "edgex/command/response/#"      # for subscribing to device service responses
+    RequestCommandTopic = "/command/request/#"     # for subscribing to internal command requests
+    ResponseCommandTopicPrefix = "/command/response"     # for publishing responses back to internal service /<device-name>/<command-name>/<method> will be added to this publish topic prefix
+    RequestQueryTopic = "/commandquery/request"   # for subscribing to internal command query requests
+    ResponseQueryTopic = "/commandquery/response"   # for publishing reponsses back to internal service
```
Comment (cloudxxx8) [inline cmd/core-command/res/configuration.toml:62]: according to your naming convention, you should either rename `RequestDeviceTopicPrefix` or `CommnadRequestTopic`

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

**Review by cloudxxx8 — APPROVED** [review]
LGTM

**Pull request merged**